<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Amplification test — amptester • chipPCR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Amplification test — amptester" />
<meta property="og:description" content="The amptester function tests if an amplification is significant." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">chipPCR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/chipPCR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/PCRuniversum/chipPCR/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Amplification test</h1>
    
    <div class="hidden name"><code>amptester.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The <code>amptester</code> function tests if an amplification is significant.</p>
    </div>

    <pre class="usage"><span class='fu'>amptester</span><span class='op'>(</span><span class='va'>y</span>, manual <span class='op'>=</span> <span class='cn'>FALSE</span>, noiselevel <span class='op'>=</span> <span class='fl'>0.08</span>, background <span class='op'>=</span> <span class='cn'>NULL</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>y</th>
      <td><p>is a vector containing the fluorescence values.</p></td>
    </tr>
    <tr>
      <th>manual</th>
      <td><p>switches between a statistical test (based on a Wilcoxon rank sum 
test (<code><a href='https://rdrr.io/r/stats/wilcox.test.html'>wilcox.test</a></code>)) or manual test for a positive 
amplification signal.</p></td>
    </tr>
    <tr>
      <th>noiselevel</th>
      <td><p>can be set to a user defined value as threshold for a significant 
	amplification signal.</p></td>
    </tr>
    <tr>
      <th>background</th>
      <td><p>is the range of the background signal in the amplification curve. 
The values can be added by the user or taken from the 
<code><a href='bg.max.html'>bg.max</a></code> function. Ignored if <code>manual</code> is
<code>TRUE</code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function tries to estimate if data indicates an amplification process. 
Several instances of tests are included. The first involves a 
semiautomatic test if the range of the background is lower than the range 
of the assumed signal. To differ between the ranges an instance of 
<code><a href='bg.max.html'>bg.max</a></code> is used. Herein, this function assumes that an 
amplification takes place in case the signal of the amplification is larger 
than the <code><a href='https://rdrr.io/r/stats/median.html'>median</a></code> + 5 * <code><a href='https://rdrr.io/r/stats/mad.html'>mad</a></code> than 
the background. The automatic test uses a Wilcoxon rank sum test 
<code><a href='https://rdrr.io/r/stats/wilcox.test.html'>wilcox.test</a></code> to compare the first and the last elements 
of the data. The input values are delivered by <code><a href='https://rdrr.io/r/utils/head.html'>head</a></code> 
and <code><a href='https://rdrr.io/r/utils/head.html'>tail</a></code>, respectively. For other methods please refer 
to the references listed below. Instead of assigning a zero to negative 
amplification reaction uses the current implementation of 
<code>amptester</code> very small random values. This is because 
some post function might fail in case all values are set to zero.</p>
<p># FIRST TEST - Shapiro test (SHt)</p>
<p>This is a simple test based on the hypothesis that in case 
amplification curve data come from noise is the distribution similar to a normal 
distribution. The Shapiro's normality test is used to test this hypothesis. If 
p is &gt;= 5e-04, then the distribution of the curve data indicates noise (no 
amplification).</p>
<p># SECOND TEST - Resids growth test (RGt)</p>
<p>This tests if fluorescence values in a linear phase are stable. Whenever no amplification 
occurs, fluorescence values quickly deviate from linear model. Their standardized
residuals will be strongly correlated with their value. For real amplification curves,
situation is much more stable. Noise (that means deviations from linear model) 
in  background do not correlate strongly with the changes in fluorescence. 
The decision is based on the threshold value (here 0.5).</p>
<p># THIRD TEST - Linear Regression test (LRt)</p>
<p>This test determines the R^2 by a linear regression. The R^2 are determined from 
a run of circa 15 percent range of the data. If a sequence of more than six R^2s 
is larger than 0.8 is found that is likely a nonlinear signal. This is a bit 
counterintuitive because R^2 of nonlinear data should be low.</p>
<p># FOURTH TEST (MANUAL) - Threshold test (THt)</p>
<p>This a commonly employed method. In the manual test one needs to define a 
fixed threshold. If the amplification curve signal exceeds the threshold than the 
amplification reaction is positive. Waring: This method will report positive 
amplification reaction if a negative amplification has a positive trend.</p>
<p># FOURTH TEST (AUTOMATIC) - Threshold test (THt)</p>
<p>Takes the first 20 percent and the last 15 percent of any input data set 
(amplification curve) and perform a Wilcoxon rank sum tests with the head (nh) 
and tail (nt). This test is recommended over the manual THt. Warning: This 
method may report positive amplification reaction if a negative amplification 
has a positive trend.</p>

<p># FIFTH TEST - Signal level test (SLt)</p>
<p>The meaningfulness of an amplification curve reaction can be tested by 
comparison of the signals 1) A robust "sigma" rule by median + 2 * mad 2) 
comparison of the signal/noise ratio. If less than 1.25 (25 percent) signal 
increase it is likely that nothing happened during the reaction. Waring: This 
method may report positive amplification reaction if a negative amplification 
has a positive trend.</p>
<p># SIXTH TEST - pco test (pco)</p>
<p>This test determines if the points in an amplification curve (like a polygon, in 
particular non-convex polygons) are in a "clockwise" order. The sum over the 
edges result in a positive value if the amplification curve is "clockwise" and 
is negative if the curve is counter-clockwise ((x2 - x1)(y2 + y1)). From 
experience is noise positive and "true" amplification curves "highly" negative. 
This test depends on the definition of a threshold.</p>
<p># SEVENTH TEST - Slope ratio (SlR)</p>
<p>Uses the inder function to find the approximated first derivative maximum, 
second derivative minimum and the second derivative maximum. Next the raw 
fluorescence at the approximated second derivative minimum and the second 
derivative maximum are taken from the original data set. The fluorescence 
intensities are normalized to the maximum fluorescence of this data. This data 
is used for a linear regression. Where the slope is used.</p>
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>An object of <code><a href='amptest.html'>amptest</a></code> class containing result of the test as well as
the original data.</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    

<p>Frank, D. N. BARCRAWL and BARTAB: software tools for the design and 
implementation of barcoded primers for highly multiplexed DNA sequencing 
<em>BMC bioinformatics</em>, 2009, Vol. 10, pp. 362</p>
<p>Peirson, S. N., Butler, J. N. and Foster, R. G. Experimental validation of 
novel and conventional approaches to quantitative real-time PCR data analysis 
<em>Nucleic Acids Research</em>, 2003, Vol. 31(14), pp. e73-e73</p>
<p>Rao, X., Lai, D. and Huang, X. A New Method for Quantitative Real-Time 
Polymerase Chain Reaction Data Analysis <em>Journal of Computational 
Biology</em>, 2013, Vol. 20(9), pp. 703-711</p>
<p>Ruijter, J. M., Ramakers, C., Hoogaars, W. M. H., Karlen, Y., Bakker, O., 
Hoff, M. J. B. v. d. and Moorman, A. F. M. Amplification efficiency: 
linking baseline and bias in the analysis of quantitative PCR data, 
<em>Nucleic Acids Research</em>, 2009, Vol. 37(6), pp. e45-e45</p>
<p>Rutledge, R. G. and Stewart, D. A kinetic-based sigmoidal model for the 
polymerase chain reaction and its application to high-capacity absolute 
quantitative real-time PCR <em>BMC biotechnology</em>, 2008, Vol. 8, pp. 47</p>
<p>Tichopad, A., Dilger, M., Schwarz, G. and Pfaffl, M. W. Standardized 
determination of real-time PCR efficiency from a single reaction set-up 
<em>Nucleic Acids Research</em>, 2003, Vol. 31(20), pp. e122</p>
<p>Wilhelm, J., Pingoud, A. and Hahn, M. SoFAR: software for fully automatic 
evaluation of real-time PCR data <em>BioTechniques</em>, 2003, Vol. 34(2), 
pp. 324-332</p>
<p>Zhao, S. and Fernald, R. D. Comprehensive Algorithm for Quantitative 
Real-Time Polymerase Chain Reaction <em>Journal of computational biology: 
a journal of computational molecular cell biology</em>, 2005, Vol. 12(8), pp. 
1047-1064</p>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Stefan Roediger, Michal Burdukiewicz</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># First example</span>
<span class='co'># Arrange graphs in orthogonal matrix and set parameter for the plot.</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/par.html'>par</a></span><span class='op'>(</span>las <span class='op'>=</span> <span class='fl'>0</span>, bty <span class='op'>=</span> <span class='st'>"n"</span>, cex.axis <span class='op'>=</span> <span class='fl'>1.5</span>, cex.lab <span class='op'>=</span> <span class='fl'>1.5</span>, 
    font <span class='op'>=</span> <span class='fl'>2</span>, cex.main <span class='op'>=</span> <span class='fl'>1.8</span>, oma <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>,<span class='fl'>1</span>,<span class='fl'>1</span>,<span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span>
    
<span class='co'># Simulation of an amplification curve with 40 cycles and a Cq of</span>
<span class='co'># circa 28. The amplification curve of "pos" (positive) has low </span>
<span class='co'># noise and the amplification curve of "neg" (negative) has high </span>
<span class='co'># noise.</span>

<span class='va'>pos</span> <span class='op'>&lt;-</span> <span class='fu'><a href='AmpSim.html'>AmpSim</a></span><span class='op'>(</span>cyc <span class='op'>=</span> <span class='fl'>1</span><span class='op'>:</span><span class='fl'>40</span>, Cq <span class='op'>=</span> <span class='fl'>28</span>, noise <span class='op'>=</span> <span class='cn'>TRUE</span>, nnl <span class='op'>=</span> <span class='fl'>0.03</span><span class='op'>)</span>
<span class='va'>neg</span> <span class='op'>&lt;-</span> <span class='fu'><a href='AmpSim.html'>AmpSim</a></span><span class='op'>(</span>cyc <span class='op'>=</span> <span class='fl'>1</span><span class='op'>:</span><span class='fl'>40</span>, Cq <span class='op'>=</span> <span class='fl'>28</span>, noise <span class='op'>=</span> <span class='cn'>TRUE</span>, nnl <span class='op'>=</span> <span class='fl'>0.8</span><span class='op'>)</span>

<span class='co'># Plot the raw data of the simulations.</span>

<span class='fu'><a href='https://rdrr.io/r/graphics/par.html'>par</a></span><span class='op'>(</span>fig <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>,<span class='fl'>0.5</span>,<span class='fl'>0.5</span>,<span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='cn'>NA</span>, <span class='cn'>NA</span>, xlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>40</span><span class='op'>)</span>, ylim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>2.1</span><span class='op'>)</span>, xlab <span class='op'>=</span> <span class='st'>"Cycles"</span>, 
     ylab <span class='op'>=</span> <span class='st'>"Fluorescence"</span>, main <span class='op'>=</span> <span class='st'>"qPCR - Raw data"</span>, type <span class='op'>=</span> <span class='st'>"b"</span><span class='op'>)</span>
</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/graphics/mtext.html'>mtext</a></span><span class='op'>(</span><span class='st'>"A"</span>, cex <span class='op'>=</span> <span class='fl'>2</span>, side <span class='op'>=</span> <span class='fl'>3</span>, adj <span class='op'>=</span> <span class='fl'>0</span>, font <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>
</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/graphics/points.html'>points</a></span><span class='op'>(</span><span class='va'>pos</span>, col <span class='op'>=</span> <span class='fl'>1</span>, typ <span class='op'>=</span> <span class='st'>"b"</span>, pch <span class='op'>=</span> <span class='fl'>19</span><span class='op'>)</span>
</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/graphics/points.html'>points</a></span><span class='op'>(</span><span class='va'>neg</span>, col <span class='op'>=</span> <span class='fl'>2</span>, typ <span class='op'>=</span> <span class='st'>"b"</span>, pch <span class='op'>=</span> <span class='fl'>20</span><span class='op'>)</span>
</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/graphics/legend.html'>legend</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>2</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Positive"</span>, <span class='st'>"Negative Control (noise)"</span><span class='op'>)</span>, 
         pch <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>19</span>,<span class='fl'>20</span><span class='op'>)</span>, col <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>,<span class='fl'>2</span><span class='op'>)</span>, lwd <span class='op'>=</span> <span class='fl'>2</span>, bty <span class='op'>=</span> <span class='st'>"n"</span><span class='op'>)</span>
</div><div class='input'>
<span class='co'># Plot data again after an analysis by ampteser. "neg" is set to small </span>
<span class='co'># random numbers, while "pos" remains unchanged.</span>

<span class='fu'><a href='https://rdrr.io/r/graphics/par.html'>par</a></span><span class='op'>(</span>fig <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>,<span class='fl'>0.5</span>,<span class='fl'>0</span>,<span class='fl'>0.5</span><span class='op'>)</span>, new <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='cn'>NA</span>, <span class='cn'>NA</span>, xlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>40</span><span class='op'>)</span>, ylim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>2.1</span><span class='op'>)</span>, xlab <span class='op'>=</span> <span class='st'>"Cycles"</span>, 
     ylab <span class='op'>=</span> <span class='st'>"Fluorescence"</span>, main <span class='op'>=</span> <span class='st'>"qPCR - amptester"</span>, type <span class='op'>=</span> <span class='st'>"b"</span><span class='op'>)</span>
</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/graphics/points.html'>points</a></span><span class='op'>(</span><span class='fu'>amptester</span><span class='op'>(</span><span class='va'>pos</span><span class='op'>[</span>, <span class='fl'>2</span><span class='op'>]</span><span class='op'>)</span>, col <span class='op'>=</span> <span class='fl'>1</span>, type <span class='op'>=</span> <span class='st'>"b"</span>, pch <span class='op'>=</span> <span class='fl'>19</span><span class='op'>)</span>
</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/graphics/points.html'>points</a></span><span class='op'>(</span><span class='fu'>amptester</span><span class='op'>(</span><span class='va'>neg</span><span class='op'>[</span>, <span class='fl'>2</span><span class='op'>]</span><span class='op'>)</span>, col <span class='op'>=</span> <span class='fl'>2</span>, type <span class='op'>=</span> <span class='st'>"b"</span>, pch <span class='op'>=</span> <span class='fl'>20</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>The distribution of the curve data indicates noise.</span>
#&gt; <span class='message'>	 The data should be visually inspected.</span></div><div class='input'><span class='fu'><a href='https://rdrr.io/r/graphics/legend.html'>legend</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>2</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Positive"</span>, <span class='st'>"Negative Control (noise)"</span><span class='op'>)</span>, 
       pch <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>19</span>,<span class='fl'>20</span><span class='op'>)</span>, col <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>,<span class='fl'>2</span><span class='op'>)</span>, lwd <span class='op'>=</span> <span class='fl'>2</span>, bty <span class='op'>=</span> <span class='st'>"n"</span><span class='op'>)</span>
</div><div class='input'>
<span class='co'># Use of amptester for time-dependent measurements. Amplification curves </span>
<span class='co'># from the capillaryPCR data set were processed in a loop. The results of </span>
<span class='co'># amptester are added to the raw data.</span>

<span class='fu'><a href='https://rdrr.io/r/graphics/par.html'>par</a></span><span class='op'>(</span>fig <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0.5</span>,<span class='fl'>1</span>,<span class='fl'>0</span>,<span class='fl'>1</span><span class='op'>)</span>, new <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='va'>colors</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/palettes.html'>rainbow</a></span><span class='op'>(</span><span class='fl'>8</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='cn'>NA</span>, <span class='cn'>NA</span>, xlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>,<span class='fl'>80</span><span class='op'>)</span>, ylim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>,<span class='fl'>1300</span><span class='op'>)</span>, xlab <span class='op'>=</span> <span class='st'>"Time [min]"</span>, 
     ylab <span class='op'>=</span> <span class='st'>"Voltage (micro V)"</span>, main <span class='op'>=</span> <span class='st'>"ccPCR"</span><span class='op'>)</span>
</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/graphics/mtext.html'>mtext</a></span><span class='op'>(</span><span class='st'>"B"</span>, cex <span class='op'>=</span> <span class='fl'>2</span>, side <span class='op'>=</span> <span class='fl'>3</span>, adj <span class='op'>=</span> <span class='fl'>0</span>, font <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>
</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>sapply</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>,<span class='fl'>3</span>,<span class='fl'>5</span>,<span class='fl'>7</span><span class='op'>)</span>, <span class='kw'>function</span><span class='op'>(</span><span class='va'>i</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='va'>xy.tmp</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/cbind.html'>cbind</a></span><span class='op'>(</span><span class='va'>capillaryPCR</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>750</span>, <span class='va'>i</span><span class='op'>]</span>, <span class='va'>capillaryPCR</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>750</span>, <span class='va'>i</span> <span class='op'>+</span> <span class='fl'>1</span><span class='op'>]</span><span class='op'>)</span>
    
<span class='co'># Use amptester to analyse the amplification curve.</span>
<span class='co'># Note: The decisions of amptester can be invoked via res.ampt@decisions</span>
<span class='co'># in the present example.</span>

    <span class='va'>res.ampt</span> <span class='op'>&lt;-</span> <span class='fu'>amptester</span><span class='op'>(</span><span class='va'>xy.tmp</span><span class='op'>[</span>, <span class='fl'>2</span><span class='op'>]</span><span class='op'>)</span>
    
<span class='co'># Use the "decisions" of amptester in a logic to automatically decide if an</span>
<span class='co'># amplification reaction is positive. In this example linear regression test</span>
<span class='co'># (lrt.test) and the threshold test (tht.dec) are used.</span>

    <span class='va'>res.ampt</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/ifelse.html'>ifelse</a></span><span class='op'>(</span><span class='va'>res.ampt</span><span class='op'>@</span><span class='va'>decisions</span><span class='op'>[</span><span class='fl'>2</span><span class='op'>]</span> <span class='op'>==</span> <span class='cn'>TRUE</span> <span class='op'>&amp;&amp;</span> 
           <span class='va'>res.ampt</span><span class='op'>@</span><span class='va'>decisions</span><span class='op'>[</span><span class='fl'>4</span><span class='op'>]</span> <span class='op'>==</span> <span class='cn'>TRUE</span>, <span class='st'>"positve"</span>, <span class='st'>"negative"</span><span class='op'>)</span>

<span class='co'># Plot the amplification curve with the decisions.</span>
    <span class='fu'><a href='https://rdrr.io/r/graphics/lines.html'>lines</a></span><span class='op'>(</span><span class='va'>xy.tmp</span><span class='op'>[</span>, <span class='fl'>1</span><span class='op'>]</span>, <span class='va'>xy.tmp</span><span class='op'>[</span>, <span class='fl'>2</span><span class='op'>]</span>, type <span class='op'>=</span> <span class='st'>"b"</span>, pch <span class='op'>=</span> <span class='fl'>20</span>, col <span class='op'>=</span> <span class='va'>colors</span><span class='op'>[</span><span class='va'>i</span><span class='op'>]</span><span class='op'>)</span>
    <span class='fu'><a href='https://rdrr.io/r/graphics/text.html'>text</a></span><span class='op'>(</span><span class='fl'>75</span>, <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>max</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/stats/na.fail.html'>na.omit</a></span><span class='op'>(</span><span class='va'>xy.tmp</span><span class='op'>[</span>, <span class='fl'>2</span><span class='op'>]</span><span class='op'>)</span><span class='op'>)</span>, <span class='va'>res.ampt</span>, cex <span class='op'>=</span> <span class='fl'>1.3</span>, col <span class='op'>=</span> <span class='va'>colors</span><span class='op'>[</span><span class='va'>i</span><span class='op'>]</span><span class='op'>)</span>
  <span class='op'>}</span>
<span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: essentially perfect fit: summary may be unreliable</span></div><div class='output co'>#&gt; <span class='warning'>Warning: essentially perfect fit: summary may be unreliable</span></div><div class='output co'>#&gt; <span class='warning'>Warning: NA values constitutes more than 0.3 of the input data.</span>
#&gt; <span class='warning'>Approximation may not be correct.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: More than 2 missing values in last 6 elements.</span>
#&gt; <span class='warning'>Approximation may not be correct.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: essentially perfect fit: summary may be unreliable</span></div><div class='output co'>#&gt; <span class='warning'>Warning: essentially perfect fit: summary may be unreliable</span></div><div class='img'><img src='amptester-1.png' alt='' width='700' height='433' /></div><div class='output co'>#&gt; [[1]]
#&gt; NULL
#&gt; 
#&gt; [[2]]
#&gt; NULL
#&gt; 
#&gt; [[3]]
#&gt; NULL
#&gt; 
#&gt; [[4]]
#&gt; NULL
#&gt; </div><div class='input'><span class='co'># Second Example</span>
<span class='co'># Example to test an amplification reaction.</span>
<span class='co'># Simulate first a positive amplification curve with 45 cycles and than a </span>
<span class='co'># negative amplification curve with 45 cycles. The negative amplification</span>
<span class='co'># curve is created from a normal distribution</span>
<span class='co'># </span>
<span class='va'>fluo.neg</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/Normal.html'>rnorm</a></span><span class='op'>(</span><span class='fl'>45</span><span class='op'>)</span>
<span class='va'>fluo.pos</span> <span class='op'>&lt;-</span> <span class='fu'><a href='AmpSim.html'>AmpSim</a></span><span class='op'>(</span>cyc <span class='op'>=</span> <span class='fl'>1</span><span class='op'>:</span><span class='fl'>45</span>, Cq <span class='op'>=</span> <span class='fl'>45</span>, ampl <span class='op'>=</span> <span class='fl'>40</span>, noise <span class='op'>=</span> <span class='cn'>TRUE</span>, 
       nnl <span class='op'>=</span> <span class='fl'>0.03</span><span class='op'>)</span><span class='op'>[</span>, <span class='fl'>2</span><span class='op'>]</span>

<span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='cn'>NA</span>, <span class='cn'>NA</span>, xlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>45</span><span class='op'>)</span>, ylim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>1</span>, <span class='fl'>45</span><span class='op'>)</span>, xlab <span class='op'>=</span> <span class='st'>"Cycles"</span>, 
     ylab <span class='op'>=</span> <span class='st'>"Fluorescence"</span>, 
     main <span class='op'>=</span> <span class='st'>"Simulation of a qPCR with 45 Cycles"</span>, type <span class='op'>=</span> <span class='st'>"b"</span><span class='op'>)</span>
</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/graphics/points.html'>points</a></span><span class='op'>(</span><span class='fu'>amptester</span><span class='op'>(</span><span class='va'>fluo.pos</span><span class='op'>)</span>, type <span class='op'>=</span> <span class='st'>"b"</span>, pch <span class='op'>=</span> <span class='fl'>20</span><span class='op'>)</span>
</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/graphics/points.html'>points</a></span><span class='op'>(</span><span class='fu'>amptester</span><span class='op'>(</span><span class='va'>fluo.neg</span><span class='op'>)</span>, type <span class='op'>=</span> <span class='st'>"b"</span>, col <span class='op'>=</span> <span class='st'>"red"</span>, pch <span class='op'>=</span> <span class='fl'>20</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>The distribution of the curve data indicates noise.</span>
#&gt; <span class='message'>	 The data should be visually inspected.</span></div><div class='input'><span class='fu'><a href='https://rdrr.io/r/graphics/points.html'>points</a></span><span class='op'>(</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>45</span>, <span class='va'>fluo.neg</span>, col <span class='op'>=</span> <span class='st'>"red"</span><span class='op'>)</span>
</div><div class='input'>
<span class='fu'><a href='https://rdrr.io/r/graphics/legend.html'>legend</a></span><span class='op'>(</span><span class='fl'>1</span>,<span class='fl'>40</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Positive"</span>, <span class='st'>"Negative Control (noise)"</span>, 
       <span class='st'>"noise pattern"</span><span class='op'>)</span>, pch <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>20</span>,<span class='fl'>20</span>,<span class='fl'>1</span><span class='op'>)</span>, col <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>,<span class='fl'>2</span>,<span class='fl'>2</span><span class='op'>)</span>, lwd <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>
</div><div class='img'><img src='amptester-2.png' alt='' width='700' height='433' /></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Stefan Roediger, Michal Burdukiewicz.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


